docs/proposals/deployment.md

```diff
@@ -0,0 +1,254 @@
+  // the start of the update (ex: 10%). This can not be 0 if MaxUnavailable is 0.
+  // Absolute number is calculated from percentage by rounding up.
+  // By default, a value of 1 is used.
+  // Example: when this is set to 30%, the new RC can be scaled up by 30%
+  // immediately when the rolling update starts. Once old pods have been killed,
+  // new RC can be scaled up further, ensuring that total number of pods running
+  // at any time during the update is atmost 130% of original pods.
+  MaxSurge IntOrString
+
+  // Minimum number of seconds for which a newly created pod should be ready
+  // without any of its container crashing, for it to be considered available.
+  // Defaults to 0 (pod will be considered available as soon as it is ready)
+  MinReadySeconds int
+}
+
+type DeploymentStatus struct {
+  // The total number of pods created with the desired template spec.
+  Replicas int
+}
+
+```
+
+### Controller
+
+#### Deployment Controller
+
+The DeploymentController will make Deployments happen.
+It will watch Deployment objects in etcd.
+For each pending deployment, it will:
+
+1. Find all RCs whose label selector is a superset of DeploymentSpec.Selector.
+   - For now, we will do this in the client - list all RCs and then filter the
+     ones we want. Eventually, we want to expose this in the API.
+2. The new RC can have the same selector as the old RC and hence we add a unique
+   selector to all these RCs (and the corresponding label to their pods) to ensure
+   that they do not select the newly created pods (or old pods get selected by
+   new RC).
+   - The label key will be "deployment.kubernetes.io/podTemplateHash".
+   - The label value will be hash of the podTemplateSpec for that RC without
+     this label. This value will be unique for all RCs, since PodTemplateSpec should be unique.
+   - If the RCs and pods dont already have this label and selector:
+     - We will first add this to RC.PodTemplateSpec.Metadata.Labels for all RCs to
+       ensure that all new pods that they create will have this label.
+     - Then we will add this label to their existing pods and then add this as a selector
+       to that RC.
+3. Find if there exists an RC for which value of "deployment.kubernetes.io/podTemplateHash" label
+   is same as hash of DeploymentSpec.PodTemplateSpec. If it exists already, then
+   this is the RC that will be ramped up. If there is no such RC, then we create
+   a new one using DeploymentSpec and then add a "deployment.kubernetes.io/podTemplateHash" label
+   to it. RCSpec.replicas = 0 for a newly created RC.
+4. Scale up the new RC and scale down the olds ones as per the DeploymentStrategy.
+   - Raise an event if we detect an error, like new pods failing to come up.
+5. Go back to step 1 unless the new RC has been ramped up to desired replicas
+   and the old RCs have been ramped down to 0.
+6. Cleanup.
+
+DeploymentController is stateless so that it can recover incase it crashes during a deployment.
```

Comment:
Hrm - when we discussed in the hangout I thought we talked about "permfail"
in a particular template if it it violates some implicit condition (like no
pods ever become ready) and we restore the original 100% availability?

On Thu, Aug 13, 2015 at 4:43 PM, Nikhil Jindal <notifications@github.com>
wrote:

> In docs/proposals/deployment.md
> <https://github.com/kubernetes/kubernetes/pull/12236#discussion_r37022839>
> :
>
> > +     - We will first add this to RC.PodTemplateSpec.Metadata.Labels for all RCs to
> > +       ensure that all new pods that they create will have this label.
> > +     - Then we will add this label to their existing pods and then add this as a selector
> > +       to that RC.
> > +3. Find if there exists an RC for which value of "deployment.kubernetes.io/podTemplateHash" label
> > +   is same as hash of DeploymentSpec.PodTemplateSpec. If it exists already, then
> > +   this is the RC that will be ramped up. If there is no such RC, then we create
> > +   a new one using DeploymentSpec and then add a "deployment.kubernetes.io/podTemplateHash" label
> > +   to it. RCSpec.replicas = 0 for a newly created RC.
> > +4. Scale up the new RC and scale down the olds ones as per the DeploymentStrategy.
> > +   - Raise an event if we detect an error, like new pods failing to come up.
> > +5. Go back to step 1 unless the new RC has been ramped up to desired replicas
> > +   and the old RCs have been ramped down to 0.
> > +6. Cleanup.
> > +
> > +DeploymentController is stateless so that it can recover incase it crashes during a deployment.
>
> I discussed this with @bgrant0607 <https://github.com/bgrant0607> and it
> doesnt look like we need conditions here.
> Deployment will be "always running" like ReplicationController. It will
> not succeed or fail, but will always keep trying. Incase of errors, we can
> add exponential backoff and generate events, but it should not keep trying.
>
> Status will have Replicas (total old+new pods) and UpdatedReplicas(new
> pods) fields which can be used to track its progress.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12236/files#r37022839>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift